keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1328 @@
+
+The `apiregistration.k8s.io` v1 APIServiceSpec object will be extended with a
+new optional field `TrustAnchors` that references a single trust anchor set
+within a single ClusterTrustBundle object.  `TrustAnchors` will be mutually
+exclusive with the existing `CABundle` and `InsecureSkipTLSVerify` fields.
+
+When `TrustAnchors` is specified, kube-apiserver will load the certificates from
+the named trust anchor set of the named ClusterTrustBundle object, an use those
+for validating the certificate presented by the aggregated API server.
+kube-apiserver will follow updates to the ClusterTrustBundle object, and use the
+updated certificates within a few seconds.
+
+```diff
+type APIServiceSpec struct {
+    // ...
+
+    // InsecureSkipTLSVerify disables TLS certificate verification when communicating with this server.
+    // This is strongly discouraged.  You should use the CABundle instead.
+    InsecureSkipTLSVerify bool `json:"insecureSkipTLSVerify,omitempty" protobuf:"varint,4,opt,name=insecureSkipTLSVerify"`
+    // CABundle is a PEM encoded CA bundle which will be used to validate an API server's serving certificate.
+    // If unspecified, system trust roots on the apiserver are used.
+    // +listType=atomic
+    // +optional
+    CABundle []byte `json:"caBundle,omitempty" protobuf:"bytes,5,opt,name=caBundle"`
++	// A reference to a dynamic trust anchor set that will be used to validate
++	// the webhook's server certificate.
++	//
++	// Mutually exclusive with CABundle and InsecureSkipTLSVerify.
++	//
++	// This field is alpha.  Using it requires setting the ClusterTrustBundle
++	// feature gate.
++	//
++	// +optional
++	TrustAnchors *TrustAnchorSetReference `json:"trustAnchors,omitempty"`
+
+    // ...
+}
++
++// TrustAnchorSetReference names a particular trust anchor set within a
++// ClusterTrustBundle object.
++type TrustAnchorSetReference struct {
++	// The ClusterTrustBundle object to refer to.
++	ClusterTrustBundleName string `json:"clusterTrustBundleName"`
++
++	// The individual named trust anchor set within the ClusterTrustBundle.
++	TrustAnchorSetName string `json:"trustAnchorSetName"`
++}
+```
+
+### trustAnchors Projected Volume Source
+
+The trustAnchors projected volume source is responsible for retrieving a
+designated trust anchor set from a signer and publishing it into the container
+filesystem.
+
+All of the new fields described in this section are gated by the
+ClusterTrustBundle kubelet feature gate.  Unless the feature gate is enabled,
+usage of these alpha objects and fields will be rejected by kubelet.
+
+#### Configuration Object Definition
+
+```go
+// TrustAnchorsProjection is the top-level configuration object for
+// instances of the trustAnchors kubelet projected volume source.
+type TrustAnchorsProjection struct {
+  // Which trust anchor sets should we retrieve from the signer, and where
+  // should we write them into the volume?
+  Files []TrustAnchorSetFile `json:"files"`
```

Comment:
This looks excessively flexible and complex. It seems like we could easily widdle this down to:

```
type TrustAnchorsProjection struct {
  // Name of ClusterTrustBundle
  Name string
  Items []corev1.KeyToPath
}
```
Or if we don't end up supporting multiple trust anchor sets:
```
type TrustAnchorsProjection struct {
  // Name of ClusterTrustBundle
  Name string
  Path string
}

Reply:
Again, it's mTLS that spoils the punch.  gRPC and envoy are building out support for a file format that will let them load trust anchors for multiple SPIFFE trust domains at the same time.  So if I operate `a.com`, and my pods will be talking to peers with certs from `b.com` and `c.com`, I can point to one file on disk that contains a key-value mapping for a.com, b.com, and c.com all together.

It seems like we should keep flexibility to eventually support formats that aren't "one single trust anchor set as a flat PEM bundle".

Reply:
I guess we could support that by calling the volume `pemTrustAnchorSet`, and then supporting any further formats as different projected volume types once some other concrete formats arrive.  That might be cleaner.

And, since we can specify a given source multiple times in a single projected volume spec, I like your further simplification.  How about
```
type PEMTrustAnchorSetProject struct {
  ClusterTrustBundleName string // Since people want namespaced TrustBundles in the future.
  Set string // *I* think this is necessary, but we can decide once we have a concrete mTLS user story to ponder.
  Path string
}
```

Reply:
Same as above, this has been significantly simplified.